### PR TITLE
fix overflows in Timestamp.tz_localize near boundaries

### LINKED ISSUE #12677: ERR: tz conversions at the min/max boundaries should fail if overflow
Timestamp comparisons with pd.Timestamp.max don't seem to work correctly when there are timezones, as in the code sample below. This is with Pandas 0.16.2.
#### Code Sample, a copy-pastable example if possible

```
import pandas as pd
import datetime

# should print True and does
print pd.Timestamp(datetime.date(2010,1,1)).tz_localize('UTC')<pd.Timestamp.max.tz_localize('UTC')
# should print True but prints False
print pd.Timestamp(datetime.date(2010,1,1)).tz_localize('US/Eastern')<pd.Timestamp.max.tz_localize('US/Eastern')
```
#### output of `pd.show_versions()`

```
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Darwin
OS-release: 15.3.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.16.2
nose: None
Cython: 0.23.4
numpy: 1.10.1
scipy: 0.16.1
statsmodels: None
IPython: 4.0.1
sphinx: None
patsy: 0.4.1
dateutil: 2.4.2
pytz: 2015.7
bottleneck: 1.0.0
numexpr: 2.4.6
matplotlib: 1.5.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
Flesh out docstring for localize_tsobject, add slightly stronger typing.

Closes #12677

### CODE REVIEW

File: pandas/_libs/tslibs/conversion.pyx [hunk 1]
```diff
@@ -490,12 +493,37 @@ cdef _TSObject convert_str_to_tsobject(object ts, object tz, object unit,
     return convert_to_tsobject(ts, tz, unit, dayfirst, yearfirst)
 
 
+cdef inline _check_silent_overflows(_TSObject obj):
```
Comment (jreback) [inline pandas/_libs/tslibs/conversion.pyx]: de-privatize. by definition not public if its not in the .pxd

File: pandas/_libs/tslibs/conversion.pyx [hunk 2]
```diff
@@ -490,12 +493,37 @@ cdef _TSObject convert_str_to_tsobject(object ts, object tz, object unit,
...
+    elif obj.dts.year == 2262:
+        if not (obj.value > 0):
+            raise OutOfBoundsDatetime
+
+
 # ----------------------------------------------------------------------
 # Localization
 
-cdef inline void _localize_tso(_TSObject obj, object tz):
+cdef inline void _localize_tso(_TSObject obj, tzinfo tz):
```
Comment (jreback) [inline pandas/_libs/tslibs/conversion.pyx]: can you de-privatize

File: pandas/tests/scalar/timestamp/test_timezones.py [hunk 3]
```diff
@@ -15,12 +15,29 @@
 import pandas.util._test_decorators as td
 
 from pandas import Timestamp, NaT
+from pandas._libs.tslib import OutOfBoundsDatetime
```
Comment (jreback) [inline pandas/tests/scalar/timestamp/test_timezones.py]: from pandas.errors

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note, small changes. ping on green.

File: pandas/_libs/tslibs/conversion.pyx [hunk 4]
```diff
@@ -309,12 +309,13 @@ cdef convert_to_tsobject(object ts, object tz, object unit,
...
-        check_dts_bounds(&obj.dts)
-
     if tz is not None:
-        _localize_tso(obj, tz)
+        localize_tso(obj, tz)
 
+    if obj.value != NPY_NAT:
+        # check_silent_overflows needs to run after localize_tso
+        check_dts_bounds(&obj.dts)
+        check_silent_overflows(obj)
```
Comment (jreback) [inline pandas/_libs/tslibs/conversion.pyx]: rename to check_overlows

File: pandas/_libs/tslibs/conversion.pyx [hunk 5]
```diff
@@ -490,12 +493,37 @@ cdef _TSObject convert_str_to_tsobject(object ts, object tz, object unit,
     return convert_to_tsobject(ts, tz, unit, dayfirst, yearfirst)
 
 
+cdef inline check_silent_overflows(_TSObject obj):
```
Comment (jreback) [inline pandas/_libs/tslibs/conversion.pyx]: rename, add doc-string (just convert your comment)

File: pandas/_libs/tslibs/conversion.pyx [hunk 6]
```diff
@@ -490,12 +493,37 @@ cdef _TSObject convert_str_to_tsobject(object ts, object tz, object unit,
...
+    representation of that timestamp in the given timezone.
+
+    Parameters
+    ----------
+    obj : _TSObject
+    tz : tzinfo
+
+    Returns
+    -------
+    (void)
```
Comment (jreback) [inline pandas/_libs/tslibs/conversion.pyx]: None

**Review by jreback — CHANGES_REQUESTED** [review]
more minor comment

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping ongreen.

**Comment by jbrockmendel** [comment]
Ping

**Comment by jreback** [comment]
thanks

**Comment by jreback** [comment]
some warnings as a result of this change

```
pandas/tests/scalar/timestamp/test_timezones.py::TestTimestampTZOperations::()::test_tz_localize_pushes_out_of_bounds
  /home/travis/build/pandas-dev/pandas/pandas/tests/scalar/timestamp/test_timezones.py:32: RuntimeWarning: overflow encountered in long_scalars
    Timestamp.min.tz_localize('Asia/Tokyo')
  /home/travis/build/pandas-dev/pandas/pandas/tests/scalar/timestamp/test_timezones.py:39: RuntimeWarning: overflow encountered in long_scalars
    Timestamp.max.tz_localize('US/Pacific')
```

**Comment by jreback** [comment]
when you have a chance can you look at the warnings above: https://github.com/pandas-dev/pandas/pull/19626#issuecomment-364762511

**Pull request merged**